src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensEditsHandler.cs

```diff
@@ -0,0 +1,212 @@
+            {
+                fullTokens.Add(new SemanticToken(tokens[i], tokens[i + 1], tokens[i + 2], tokens[i + 3], tokens[i + 4]));
+            }
+
+            return fullTokens.ToArray();
+        }
+
+        internal static LSP.SemanticTokensEdit GenerateEdit(int start, int deleteCount, int[] data)
+            => new LSP.SemanticTokensEdit
+            {
+                Start = start,
+                DeleteCount = deleteCount,
+                Data = data
+            };
+
+        private sealed class LongestCommonSemanticTokensSubsequence : LongestCommonSubsequence<SemanticToken[]>
+        {
+            private static readonly LongestCommonSemanticTokensSubsequence s_instance = new LongestCommonSemanticTokensSubsequence();
+
+            protected override bool ItemsEqual(
+                SemanticToken[] oldSemanticTokens, int oldIndex,
+                SemanticToken[] newSemanticTokens, int newIndex)
+            {
+                var oldToken = oldSemanticTokens[oldIndex];
+                var newToken = newSemanticTokens[newIndex];
+
+                return oldToken.DeltaLine == newToken.DeltaLine && oldToken.DeltaStartCharacter == newToken.DeltaStartCharacter &&
+                    oldToken.Length == newToken.Length && oldToken.TokenType == newToken.TokenType &&
+                    oldToken.TokenModifiers == newToken.TokenModifiers;
+            }
+
+            // We reverse the result array since the original edits are reported with last tokens in the
+            // document output first in the IEnumerable. This is due to the nature of the Myers diff
+            // algorithm in which the optimal steps are retraced from the goal. However, when creating
+            // SemanticTokensEdits to report back to LSP, we need to analyze the edits in chronological order.
+            public static IEnumerable<SequenceEdit> GetEdits(
+                SemanticToken[] oldSemanticTokens, SemanticToken[] newSemanticTokens)
+                => s_instance.GetEdits(oldSemanticTokens, oldSemanticTokens.Length, newSemanticTokens, newSemanticTokens.Length).Reverse();
+        }
+
+        /// <summary>
+        /// Stores the values that make up the LSP representation of an individual semantic token.
+        /// </summary>
+        private struct SemanticToken
+        {
+            public int DeltaLine { get; }
+            public int DeltaStartCharacter { get; }
+            public int Length { get; }
+            public int TokenType { get; }
+            public int TokenModifiers { get; }
+
+            public SemanticToken(int deltaLine, int deltaStartCharacter, int length, int tokenType, int tokenModifiers)
+            {
+                DeltaLine = deltaLine;
+                DeltaStartCharacter = deltaStartCharacter;
+                Length = length;
+                TokenType = tokenType;
+                TokenModifiers = tokenModifiers;
+            }
+
+            public int[] ToArray()
```

Comment:
don't do this.  this will allocate an array for every semantic token.  instead, just splat this into place in teh final array (many ways to do this).

Reply:
(note: this may not apply.  i see how you're using ToArray now.

Reply:
Yes, I'm not sure if it applies since it's just converting our custom SemanticToken struct type back to it's original int array format.